### PR TITLE
SI-5734 Allow setting of socket timeout for remote actors

### PR DESCRIPTION
**Pull request by martinmcnulty**
TcpService is locked while connections are being established, preventing
messages being sent to remote nodes. There was no way to specify a connect
timeout, so the service was waiting for a ConnectException to be thrown if
the remote node was unavailable (which can take several minutes).

Added a system property to allow setting a socket connect timeout to
prevent the TcpService from being locked for minutes at a time.

Default behaviour is unchanged, all tests pass and the patch has been
running in production without issue.

### CODE REVIEW

**Comment by adriaanm** [comment]
review by @phaller

**Comment by adriaanm** [comment]
@martinmcnulty, please sign the Scala CLA by logging in with your github account at http://typesafe.com/contribute/cla/scala

**Comment by martinmcnulty** [comment]
Chasing my employer on this now - hopefully I'll get an answer out of them tomorrow!  Thanks

**Comment by adriaanm** [comment]
Thank you. Sorry about the overhead.

**Comment by martinmcnulty** [comment]
No problem - thanks for being patient with my multiple attempts to get this right!  

I've now signed the CLA.

**Comment by adriaanm** [comment]
Great, thanks! @phaller, over to you

File: src/actors/scala/actors/remote/TcpService.scala [hunk 1]
```diff
@@ -59,6 +59,23 @@ object TcpService {
     portnum
   }
 
+  val connectTimeoutMillis = {
```
Comment (phaller) [inline src/actors/scala/actors/remote/TcpService.scala]: `private`?

File: src/actors/scala/actors/remote/TcpService.scala [hunk 2]
```diff
@@ -176,7 +193,15 @@ class TcpService(port: Int, cl: ClassLoader) extends Thread with Service {
   }
 
   def connect(n: Node): TcpServiceWorker = synchronized {
-    val socket = new Socket(n.address, n.port)
+    val socket = new Socket()
+    val start = System.nanoTime()
```
Comment (phaller) [inline src/actors/scala/actors/remote/TcpService.scala]: I'd drop the `()` for consistency with the other use below

File: src/actors/scala/actors/remote/TcpService.scala [hunk 3]
```diff
@@ -176,7 +193,15 @@ class TcpService(port: Int, cl: ClassLoader) extends Thread with Service {
...
+    try {
+      socket.connect(new InetSocketAddress(n.address, n.port), TcpService.connectTimeoutMillis)
+    } catch {
+      case e: SocketTimeoutException =>
+        Debug.warning(f"Timed out connecting to [${n.address}%s:${n.port}%s] after ${(System.nanoTime - start) / math.pow(10, 9)}%.3f seconds")
```
Comment (phaller) [inline src/actors/scala/actors/remote/TcpService.scala]: Should we use the default `toString` of case class `Node` for consistency?

File: src/actors/scala/actors/remote/TcpService.scala (hunk 2 above)
Comment (martinmcnulty) [inline src/actors/scala/actors/remote/TcpService.scala]: Sure - for the Socket too or just the nanoTime call?

File: src/actors/scala/actors/remote/TcpService.scala (hunk 1 above)
Comment (martinmcnulty) [inline src/actors/scala/actors/remote/TcpService.scala]: Good plan

File: src/actors/scala/actors/remote/TcpService.scala (hunk 3 above)
Comment (martinmcnulty) [inline src/actors/scala/actors/remote/TcpService.scala]: Sounds sensible - I'll commit this now

**Comment by martinmcnulty** [comment]
@phaller - I think I've made all the changes you suggested. Anything else I need to do?

Thanks

**Comment by phaller** [comment]
LGTM

**Pull request merged**